tools/krb5-sync.8
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "KRB5-SYNC 8"
.TH KRB5-SYNC 8 "2015-08-19" "3.1" "krb5-sync"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
krb5\-sync \- Synchronize passwords and status with Active Directory
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBkrb5\-sync\fR [\fB\-d\fR | \fB\-e\fR] [\fB\-p\fR \fIpassword\fR] \fIuser\fR
.PP
\&\fBkrb5\-sync\fR \fB\-f\fR \fIfile\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBkrb5\-sync\fR provides a command-line interface to the same functions
provided by the password and status synchronization plugin.  It can push a
new password to Active Directory (actually, to any password store that
supports the Kerberos set-password protocol) or activate or deactivate an
account in Active Directory.
.PP
To synchronize passwords, provide the \fB\-p\fR option and specify the
password.  Note that the password is given on the command line and must be
quoted if it contains special characters, and the password will be exposed
to any other users on the system where this command is run.  This is
useful primarily for testing and should not be used with production
passwords.  Synchronization to Active Directory will be attempted based on
the configuration in \fIkrb5.conf\fR (see below).
.PP
To enable or disable an account, provide the \fB\-e\fR or \fB\-d\fR option
respectively.  These options can also be provided in conjunction with the
\&\fB\-p\fR option to take both actions at once.
.PP
In either case, \fIuser\fR should be the principal name for which these
actions should be taken.  \fIuser\fR may be either unqualified or in the
local realm; either way, the Active Directory realm in which to make
changes will be taken from the \fIkrb5.conf\fR configuration.
.PP
Alternately, \fBkrb5\-sync\fR also supports processing actions from a file.
To do this, use the \fB\-f\fR flag and give the file on the command line.  The
format of the file should be as follows:
.PP
.Vb 4
\&    <account>
\&    ad
\&    password | enable | disable
\&    <password>
.Ve
.PP
where the fourth line is present only if the <action> is \f(CW\*(C`password\*(C'\fR.
<account> should be the unqualified name of the account.  The second line
should be the string \f(CW\*(C`ad\*(C'\fR to push the change to Windows Active Directory.
The third line should be one of \f(CW\*(C`password\*(C'\fR, \f(CW\*(C`enable\*(C'\fR, or \f(CW\*(C`disable\*(C'\fR,
corresponding to the \fB\-p\fR, \fB\-e\fR, and \fB\-d\fR options respectively.  The
\&\f(CW\*(C`enable\*(C'\fR and \f(CW\*(C`disable\*(C'\fR actions are only supported for \s-1AD.\s0
.PP
The file format is not particularly forgiving.  In particular, all of the
keywords are case-sensitive and there must not be any whitespace at the
beginning or end of the lines (except in the password, and only if that
whitespace is part of the password), just a single newline terminating
each line.
.PP
When the \fB\-f\fR option is given, the file will be deleted if the action was
successful but left alone if the action failed.
.PP
The configuration block in \fIkrb5.conf\fR should look something like this:
.PP
.Vb 7
\&    krb5\-sync = {
\&        ad_keytab       = /etc/krb5kdc/ad\-keytab
\&        ad_principal    = service/sync@WINDOWS.EXAMPLE.COM
\&        ad_realm        = WINDOWS.EXAMPLE.COM
\&        ad_admin_server = dc1.windows.example.com
\&        ad_ldap_base    = ou=People
\&    }
.Ve
.PP
If the configuration required for an action is not given, that action will
not be performed but will apparently succeed from the perspective of the
\&\fBkrb5\-sync\fR utility.  Therefore, if this utility reports success but no
change is happening, double-check the configuration to ensure that all
required options are present.
.PP
The \f(CW\*(C`ad_keytab\*(C'\fR option specifies the location of a keytab for
authenticating to the other realm, the \f(CW\*(C`ad_principal\*(C'\fR option specifies
the principal to authenticate as (using the key in the keytab), and the
\&\f(CW\*(C`ad_realm\*(C'\fR option specifies the foreign realm.  \f(CW\*(C`ad_admin_server\*(C'\fR is the
host to contact via \s-1LDAP\s0 to push account status changes.  \f(CW\*(C`ad_ldap_base\*(C'\fR
specifies the base tree inside Active Directory where account information
is stored.  Omit the trailing \f(CW\*(C`dc=\*(C'\fR part; it will be added automatically
from \f(CW\*(C`ad_realm\*(C'\fR.
.PP
Be aware that the \f(CW\*(C`ad_instances\*(C'\fR, \f(CW\*(C`ad_base_instance\*(C'\fR, and
\&\f(CW\*(C`ad_queue_only\*(C'\fR configuration options that are used by the krb5\-sync
plugin are ignored by \fBkrb5\-sync\fR.  The \fBkrb5\-sync\fR command will push
changes to whatever principal it was given, regardless of the normal
limits on instances, and does not do any of the principal remapping
configured with \f(CW\*(C`ad_base_instance\*(C'\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-d\fR" 4
.IX Item "-d"
Disable the specified user in Active Directory.  Requires that all of the
ad_* options be set in \fIkrb5.conf\fR.  This option may not be specified at
the same time as \fB\-e\fR.
.IP "\fB\-e\fR" 4
.IX Item "-e"
Enable the specified user in Active Directory.  Requires that all of the
ad_* options be set in \fIkrb5.conf\fR.  This option may not be specified at
the same time as \fB\-e\fR.
.IP "\fB\-f\fR \fIfile\fR" 4
.IX Item "-f file"
Rather than perform a particular action based on a username given on the
command line, read a queue file and take action based on it.  The format
of the queue file is described above.  If the action fails, the file will
be left alone.  If the action succeeds, the file will be deleted.
.IP "\fB\-p\fR \fIpassword\fR" 4
.IX Item "-p password"
Change the user's password to \fIpassword\fR in Active Directory.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
Disable the account \*(L"jdoe\*(R" in Active Directory (using the \s-1AD\s0 configuration
found in \fIkrb5.conf\fR):
.PP
.Vb 1
\&    krb5\-sync \-d jdoe
.Ve
.PP
Change the password of the account \f(CW\*(C`testuser\*(C'\fR in Active Directory to
\&\f(CW\*(C`changeme\*(C'\fR:
.PP
.Vb 1
\&    krb5\-sync \-p changeme testuser@EXAMPLE.COM
.Ve
.PP
The same, except also enable the account in Active Directory:
.PP
.Vb 1
\&    krb5\-sync \-e \-p changeme testuser
.Ve
.PP
Note that the realm for the user given on the command line is optional and
ignored.
.PP
Given a file named \fIjdoe\-ad\-1168560492\fR containing:
.PP
.Vb 4
\&    jdoe
\&    ad
\&    password
\&    changeme
.Ve
.PP
the command:
.PP
.Vb 1
\&    krb5\-sync \-f jdoe\-ad\-1168560492
.Ve
.PP
will change jdoe's password to \f(CW\*(C`changeme\*(C'\fR in Active Directory and then
delete the file.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
The current version of this program is available from its web page at
<http://www.eyrie.org/~eagle/software/krb5\-sync/>.
.SH "AUTHOR"
.IX Header "AUTHOR"
Russ Allbery <eagle@eyrie.org>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright 2007, 2008, 2010, 2012 The Board of Trustees of the Leland
Stanford Junior University
.PP
Copying and distribution of this file, with or without modification, are
permitted in any medium without royalty provided the copyright notice and
this notice are preserved.  This file is offered as-is, without any
warranty.


NEWS
                      User-Visible krb5-sync Changes

krb5-sync 3.1 (2015-08-18)

    In krb5-sync-backend silent mode, fix the ignore regex for missing
    users and report unsuppressed errors properly with an ending newline.

    Update to rra-c-util 5.8:

    * Support the Solaris 10 embedded Kerberos implementation.
    * Use calloc or reallocarray instead of malloc.
    * Fix compilation with a C++ compiler.

    Update to C TAP Harness 3.4:

    * Display verbose test results with -v or C_TAP_VERBOSE.
    * Reopen standard input to /dev/null when running a test list.
    * Don't leak extraneous file descriptors to tests.

krb5-sync 3.0 (2013-12-09)

    The default installed module name has been changed to sync.so from
    krb5_sync.so, since the krb5 part is redundant in the Kerberos plugin
    context.  This will require configuration changes for existing users
    to load the new plugin path name.

    The meaning of the ad_ldap_base configuration option has changed, and
    it's now mandatory for status synchronization.  This setting should
    now contain the full DN of the tree in Active Directory where account
    information is stored (such as cn=Accounts,dc=example,dc=com).
    Previously, the dc components should be omitted and were derived from
    the realm; this is no longer done.  If this configuration option is
    not set, principal status will not be synchronized to Active
    Directory.

    Drop support for MIT Kerberos versions prior to 1.9.  All major
    distributions are now shipping with a newer version of MIT Kerberos
    than this, and supporting older versions requires supporting patches
    and maintaining handicapped internal APIs.  MIT Kerberos 1.9 and later
    do not require patches to use this module.  Patches for Heimdal are
    still provided.

    Add a new string krb5.conf option, ad_base_instance, which, if set,
    changes the way that password synchronization is handled.  When this
    option is set, the password for the principal formed by appending that
    instance to a base principal is propagated to Active Directory as the
    password for the base principal.  For example, if this is set to the
    string "windows", the password of the principal "user/windows" is
    propagated to Active Directory as the password for the principal
    "user" and password changes for the principal "user" are ignored.
    This special behavior only happens if "user/windows" exists in the
    local Kerberos KDC database; if not, password propagation for the
    principal "user" happens normally, just as if this option weren't set.
    This allows the Active Directory principal to be treated as an
    instance rather than a main account for specific users without
    affecting behavior for other users.

    Add a new boolean krb5.conf option, ad_queue_only, which, if set to
    true, forces all changes to be queued even if there are no conflicting
    changes already queued.  The changes can then be processed later with
    krb5-sync-backend.  This can be useful if real-time updates to Active
    Directory cause performance issues in kadmind or kpasswdd.  kpasswd
    clients in particular are often intolerant of delays.

    Add a new boolean krb5.conf option, syslog, which can be set to false
    to suppress syslog logging of the actions taken by the plugin and
    error messages leading to queuing the change.  Always log the error
    that leads to queuing a status change.

    Any time an Active Directory password change fails, queue the change
    instead of failing it, rather than trying to distinguish between local
    errors that should fail the change and errors that should be queued.
    The previous logic was very Stanford-specific.

    krb5-sync-backend now requires the IPC::Run and Net::Remctl::Backend
    modules be installed.  The former is available from CPAN, and the
    latter is available from the remctl package (version 3.4 or later).

    krb5-sync-backend supports a new flag, -d, which specifies the
    location of the queue directory, changing the default of
    /var/spool/krb5-sync.

    When processing events in krb5-sync-backend, skip event files which no
    longer exist by the time we get to them.  This makes krb5-sync-backend
    more robust against multiple copies running at the same time.

    Update to rra-c-util 4.12:

    * Better error messages from xasprintf on failure to format output.
    * Check return status of vsnprintf properly.
    * Significant improvements to POD tests.
    * Avoid leaking a dummy symbol from the portability layer.
    * Probe for Kerberos headers with file existence checks.

    Update to C TAP Harness 2.3:

    * runtests now treats the command line as a list of tests by default.
    * The full test executable path can now be passed to runtests -o.
    * Improved harness output for tests with lazy plans.
    * Improved harness output to a terminal for some abort cases.
    * Flush harness output after each test even when not on a terminal.
    * bail and sysbail now exit with status 255 to match Test::More.
    * Suppress lazy plans and test summaries if the test failed with bail.
    * Add warn_unused_result gcc attributes to relevant functions.

krb5-sync 2.3 (2012-09-18)

    When handling password changes from MIT Kerberos, quietly ignore
    changes where the password is NULL.  These are key randomizations,
    such as from addprinc -randkey, which this module inherently can't do
    anything with.

    The plugin is now installed in a kadm5_hook subdirectory under
    krb5/plugins (under libdir in turn), matching the plugin layout used
    by MIT Kerberos.

    When krb5-sync-backend is running in silent mode, ignore "Operation
    not permitted" errors from krb5_set_password.  Heimdal 1.5.2 returns
    this error from Active Directory when attempting to change the
    password of an account that does not exist.

    Properly pass Kerberos preprocessor flags to the compiler when

tools/krb5-sync.c
/*
 * Command-line access to the krb5-sync kadmind plugin.
 *
 * This program provides command-line access to the functionality of the
 * krb5-sync kadmind plugin.  Using it, one can push password changes or
 * enabled/disabled status according to the same configuration used by the
 * plugin.  It's primarily intended for testing, but can also be used to
 * synchronize changes when the plugin previously failed for some reason.
 *
 * Written by Russ Allbery <eagle@eyrie.org>
 * Based on code developed by Derrick Brashear and Ken Hornstein of Sine
 *     Nomine Associates, on behalf of Stanford University
 * Copyright 2006, 2007, 2010, 2012, 2013
 *     The Board of Trustees of the Leland Stanford Junior University
 *
 * See LICENSE for licensing terms.
 */

#include <config.h>
#include <portable/krb5.h>
#include <portable/system.h>

#include <errno.h>
#include <syslog.h>

#include <plugin/internal.h>
#include <util/messages-krb5.h>
#include <util/messages.h>


/*
 * Change a password in Active Directory.  Print a success message if we were
 * successful, and exit with an error message if we weren't.
 */
static void
ad_password(kadm5_hook_modinfo *config, krb5_context ctx,
            krb5_principal principal, char *password, const char *user)
{
    krb5_error_code code;

    code = sync_ad_chpass(config, ctx, principal, password);
    if (code != 0)
        die_krb5(ctx, code, "AD password change for %s failed", user);
    notice("AD password change for %s succeeded", user);
}


/*
 * Change the account status in Active Directory.  Print a success message if
 * we were successful, and exit with an error message if we weren't.
 */
static void
ad_status(kadm5_hook_modinfo *config, krb5_context ctx,
          krb5_principal principal, bool enable, const char *user)
{
    krb5_error_code code;

    code = sync_ad_status(config, ctx, principal, enable);
    if (code != 0)
        die_krb5(ctx, code, "AD status change for %s failed", user);
    notice("AD status change for %s succeeded", user);
}


/*
 * Read a line from a queue file, making sure we got a complete line and
 * cutting off the trailing newline.  Doesn't return on error.
 */
static void
read_line(FILE *file, const char *filename, char *buffer, size_t bufsiz)
{
    if (fgets(buffer, bufsiz, file) == NULL)
        sysdie("cannot read from queue file %s", filename);
    if (buffer[strlen(buffer) - 1] != '\n')
        die("line too long in queue file %s", filename);
    buffer[strlen(buffer) - 1] = '\0';
}


/*
 * Read a queue file and take appropriate action based on its contents.  The
 * format is:
 *
 *     <principal>
 *     ad
 *     enable | disable | password
 *     [<password>]
 *
 * The actions are the same as from the command-line switches, except that
 * passwords are changed separately in AFS and AD.  enable and disable are not
 * supported for AFS.
 */
static void
process_queue_file(kadm5_hook_modinfo *config, krb5_context ctx,
                   const char *filename)
{
    FILE *queue;
    char buffer[BUFSIZ];
    char *user;
    krb5_principal principal;
    krb5_error_code ret;
    bool ad = false;
    bool enable = false;
    bool disable = false;
    bool password = false;

    /* Open the queue file. */
    queue = fopen(filename, "r");
    if (queue == NULL)
        sysdie("cannot open queue file %s", filename);

    /* Get user and convert into a principal. */
    read_line(queue, filename, buffer, sizeof(buffer));
    user = strdup(buffer);
    ret = krb5_parse_name(ctx, buffer, &principal);
    if (ret != 0)
        die_krb5(ctx, ret, "cannot parse user %s into principal", buffer);

    /* Get function. */
    read_line(queue, filename, buffer, sizeof(buffer));
    if (strcmp(buffer, "ad") == 0)
        ad = true;
    else
        die("unknown target system %s in queue file %s", buffer, filename);
    read_line(queue, filename, buffer, sizeof(buffer));
    if (strcmp(buffer, "enable") == 0)
        enable = true;
    else if (strcmp(buffer, "disable") == 0)
        disable = true;
    else if (strcmp(buffer, "password") == 0)
        password = true;
    else
        die("unknown action %s in queue file %s", buffer, filename);

    /* Perform the appropriate action. */
    if (password) {
        read_line(queue, filename, buffer, sizeof(buffer));
        if (ad)
            ad_password(config, ctx, principal, buffer, user);
    } else if (enable || disable) {
        ad_status(config, ctx, principal, enable, user);
    }

    /* If we got here, we were successful.  Close the file and delete it. */
    fclose(queue);
    if (unlink(filename) != 0)
        sysdie("unable to unlink queue file %s", filename);
    free(user);
}


int
main(int argc, char *argv[])
{
    int option;
    int enable = false;
    int disable = false;
    char *password = NULL;
    char *filename = NULL;
    char *user;
    kadm5_hook_modinfo *config;
    krb5_context ctx;
    krb5_error_code code;
    krb5_principal principal;

    /*
     * Actions should be logged to LOG_AUTH to go to the same place as the
     * logs from kadmind for easier log analysis.
     */
    openlog("krb5-sync", LOG_PID, LOG_AUTH);
    message_program_name = "krb5-sync";

    /* Parse command-line options. */
    while ((option = getopt(argc, argv, "def:p:")) != EOF) {
        switch (option) {
        case 'd': disable = true;       break;
        case 'e': enable = true;        break;
        case 'f': filename = optarg;    break;
        case 'p': password = optarg;    break;

        default:
            fprintf(stderr, "Usage: krb5-sync [-d | -e] [-p <pass>] <user>\n");
            exit(1);
        }
    }
    argc -= optind;
    argv += optind;
    if (argc != 1 && filename == NULL) {
        fprintf(stderr, "Usage: krb5-sync [-d | -e] [-p <pass>] <user>\n");
        exit(1);
    }
    if (argc != 0 && filename != NULL) {
        fprintf(stderr, "Usage: krb5-sync -f <file>\n");
        exit(1);
    }
    user = argv[0];
    if (enable && disable)
        die("cannot specify both -d and -e");
    if (!enable && !disable && password == NULL && filename == NULL)
        die("no action specified");
    if (filename != NULL && (enable || disable || password != NULL))
        die("must specify queue file or action, not both");

    /* Create a Kerberos context for plugin initialization. */
    code = krb5_init_context(&ctx);
    if (code != 0)
        die_krb5(ctx, code, "cannot initialize Kerberos context");

    /* Initialize the plugin. */
    code = sync_init(ctx, &config);
    if (code != 0)
        die_krb5(ctx, code, "plugin initialization failed");

    /* Now, do whatever we were supposed to do. */
    if (filename != NULL)
        process_queue_file(config, ctx, filename);
    else {
        code = krb5_parse_name(ctx, user, &principal);
        if (code != 0)
            die_krb5(ctx, code, "cannot parse user %s into principal", user);
        if (password != NULL)
            ad_password(config, ctx, principal, password, user);
        if (enable || disable)
            ad_status(config, ctx, principal, enable, user);
    }
    exit(0);
}
